cni/network/network.go

```diff
@@ -457,199 +454,243 @@ func (plugin *netPlugin) Add(args *cniSkel.CmdArgs) error {
+	 * We can delete this if statement once they fix it.
+	 * Issue link: https://github.com/kubernetes/kubernetes/issues/57253
+	 */
 
-			if resultConsAdd != nil {
-				result = resultConsAdd
-				return nil
-			}
-		}
-	}
+	if nwInfoErr == nil {
+		log.Printf("[cni-net] Found network %v with subnet %v.", networkID, nwInfo.Subnets[0].Prefix.String())
+		nwInfo.IPAMType = nwCfg.Ipam.Type
+		options = nwInfo.Options
 
-	switch nwCfg.Ipam.Type {
-	case network.AzureCNS:
-		plugin.ipamInvoker, err = NewCNSInvoker(k8sPodName, k8sNamespace)
+		result, err = plugin.handleConsecutiveAdd(args, endpointId, networkID, &nwInfo, nwCfg)
 		if err != nil {
-			log.Printf("[cni-net] Creating network %v, failed with err %v", networkId, err)
+			log.Printf("handleConsecutiveAdd failed with error %v", err)
 			return err
 		}
-	default:
-		plugin.ipamInvoker = NewAzureIpamInvoker(plugin, &nwInfo)
-	}
 
-	options := make(map[string]interface{})
-
-	if nwInfoErr != nil {
-		// Network does not exist.
-		log.Printf("[cni-net] Creating network %v.", networkId)
+		if result != nil {
+			return nil
+		}
+	}
 
-		if !nwCfg.MultiTenancy {
-			result, resultV6, err = plugin.ipamInvoker.Add(nwCfg, args, &subnetPrefix, options)
+	// Initialize azureipam/cns ipam
+	if plugin.ipamInvoker == nil {
+		switch nwCfg.Ipam.Type {
+		case network.AzureCNS:
+			plugin.ipamInvoker, err = NewCNSInvoker(k8sPodName, k8sNamespace)
 			if err != nil {
+				log.Printf("[cni-net] Creating network %v, failed with err %v", networkID, err)
 				return err
 			}
-
-			defer func() {
-				if err != nil {
-					if result != nil && len(result.IPs) > 0 {
-						if er := plugin.ipamInvoker.Delete(&result.IPs[0].Address, nwCfg, args, options); er != nil {
-							err = plugin.Errorf("Failed to cleanup when NwInfo was not nil with error %v, after Add failed with error %w", er, err)
-						}
-					}
-					if resultV6 != nil && len(resultV6.IPs) > 0 {
-						if er := plugin.ipamInvoker.Delete(&resultV6.IPs[0].Address, nwCfg, args, options); er != nil {
-							err = plugin.Errorf("Failed to cleanup when NwInfo was not nil with error %v, after Add failed with error %w", er, err)
-						}
-					}
-				}
-			}()
-		}
-
-		gateway := result.IPs[0].Gateway
-		subnetPrefix.IP = subnetPrefix.IP.Mask(subnetPrefix.Mask)
-		nwCfg.Ipam.Subnet = subnetPrefix.String()
-		// Find the master interface.
-		masterIfName := plugin.findMasterInterface(nwCfg, &subnetPrefix)
-		if masterIfName == "" {
-			err = plugin.Errorf("Failed to find the master interface")
-			return err
+		default:
+			plugin.ipamInvoker = NewAzureIpamInvoker(plugin, &nwInfo)
 		}
-		log.Printf("[cni-net] Found master interface %v.", masterIfName)
+	}
 
-		// Add the master as an external interface.
-		err = plugin.nm.AddExternalInterface(masterIfName, subnetPrefix.String())
+	// Allocate from azure ipam
+	if !nwCfg.MultiTenancy {
+		result, resultV6, err = plugin.ipamInvoker.Add(nwCfg, args, &subnetPrefix, options)
```

Comment:
wondering if it's possible to combine this in GetMultiTenancyCNIResult, the ipam is also invoked when EnableInfraVnet is enabled in that flow